lib/elixir/lib/enum.ex

```diff
@@ -717,30 +717,64 @@ defmodule Enum do
   """
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
+  def fetch([], index) when is_integer(index),
+    do: :error
+  def fetch([elem], 0),
+    do: {:ok, elem}
+  def fetch([elem], -1),
+    do: {:ok, elem}
+  def fetch([_], index) when is_integer(index),
+    do: :error
+
+  def fetch(enumerable, index)
+      when is_map(enumerable) and map_size(enumerable) == 0 and is_integer(index),
+    do: :error
+  def fetch(enumerable, 0) when is_map(enumerable) and map_size(enumerable) == 1,
+    do: :error
+  def fetch(enumerable, -1) when is_map(enumerable) and map_size(enumerable) == 1,
+    do: :error
+  def fetch(enumerable, index)
+      when is_map(enumerable) and is_integer(index) and map_size(enumerable) == 1,
+    do: :error
+
+  def fetch(enumerable, index) when is_integer(index) and index < 0 do
+    case count(enumerable) + index do
+      i when i < 0 ->
+        :error
+      i ->
+        fetch(enumerable, i)
+    end
+  end
+
+  def fetch(enumerable, index)
       when is_list(enumerable) and is_integer(index) and index >= 0 do
     do_fetch(enumerable, index)
   end
 
   def fetch(enumerable, index) when is_integer(index) and index >= 0 do
-    res =
-      Enumerable.reduce(enumerable, {:cont, 0}, fn(entry, acc) ->
-        if acc == index do
-          {:halt, entry}
-        else
-          {:cont, acc + 1}
+    # this avoids unnecessarily traversing an enumerable that implements Enumerable.count/1
+    # given an invalid index
+    with {:ok, count} when (count - 1 - index) < 0 <- Enumerable.count(enumerable)
```

Comment:
`Enumerable.count` is going to return `{:error, module}`. You should do `module.reduce` below instead of `Enumerable.reduce` to avoid another protocol lookup. Therefore I suggest using a `case` here. We didn't do this for `take_random` because we were not calling `Enumerable.reduce` directly in there.